sklego/meta/grouped_predictor.py

```diff
@@ -397,15 +208,160 @@ def decision_function(self, X):
+
+                if _y.isnull().all():
+                    estimators_[grp_values] = clone(self.estimator).fit(_X)  # TODO: or .fit(_X, None) ?
+                else:
+                    estimators_[grp_values] = clone(self.estimator).fit(_X, _y)
+
+        return estimators_
+
+    def __fit_shrinkage_factors(self, frame):
+        """Computes the shrinkage coefficients for all fitted levels (corresponding to the keys of self.estimators_)"""
+
+        check_is_fitted(self, ["estimators_", "groups_"])
+        counts = frame.groupby(self.groups_).size().rename("counts")
+        all_grp_values = list(self.estimators_.keys())
+
+        hierarchical_counts = {
+            grp_value: [counts.loc[subgroup].sum() for subgroup in expanding_list(grp_value, tuple)]
+            for grp_value in all_grp_values
+        }
+
+        shrinkage_factors = {
+            grp_value: self.shrinkage_function_(counts, **self.shrinkage_kwargs)
+            for grp_value, counts in hierarchical_counts.items()
+        }
+
+        # Normalize and pad
+        return {
+            grp_value: np.pad(shrink_array / shrink_array.sum(), (0, max(0, self.n_levels_ - shrink_array.size)))
+            for grp_value, shrink_array in shrinkage_factors.items()
+        }
+
+    def __set_shrinkage_function(self):
+        """Set the shrinkage function and validate it if it is a custom callable"""
+
+        if self.shrinkage and len(as_list(self.groups)) == 1 and not self.use_global_model:
+            raise ValueError("Cannot do shrinkage with a single group and `use_global_model=False`")
+
+        if self.shrinkage in self._ALLOWED_SHRINKAGE.keys():
+            shrinkage_function_ = self._ALLOWED_SHRINKAGE[self.shrinkage]
+
+        elif callable(self.shrinkage):
+            self.__check_shrinkage_func()
+            shrinkage_function_ = self.shrinkage
+
+        elif self.shrinkage is None:
+            shrinkage_function_ = lambda x: np.pad(np.zeros(self.n_levels_ - 1), pad_width=(0, 1), constant_values=1)
+
+        else:
+            raise ValueError(f"`shrinkage` should be either `None`, {self._ALLOWED_SHRINKAGE.keys()}, or a callable")
+        return shrinkage_function_
+
+    def __check_shrinkage_func(self):
+        """Validate the shrinkage function if a function is specified"""
+        group_lengths = [10, 5, 2]
+        expected_shape = np.array(group_lengths).shape
+        try:
+            result = self.shrinkage(group_lengths)
+        except Exception as e:
+            raise ValueError(f"Caught an exception while checking the shrinkage function: {str(e)}") from e
+        else:
+            if not isinstance(result, np.ndarray):
+                raise ValueError(f"shrinkage_function({group_lengths}) should return an np.ndarray")
+            if result.shape != expected_shape:
+                raise ValueError(f"shrinkage_function({group_lengths}).shape should be {expected_shape}")
+
+    def __predict_estimators(self, X, method_name):
```

Comment:
This is a bit of a headache but allows for a unified approach:

- `preds` is a 3d array of `(n_samples, n_levels, n_classes)` where for regression or od `n_classes` can be considered to be 1. This will be populated with a prediction for each level
- `shrinkage` is a 2d array `(n_samples, n_levels)` as shrinkage factor is already one per level. This will get the shrinkage to use for each sample. Since we are doing predictions from outer to inner most level, we are overwriting when needed.
- The `_get_estimator` returning the level allows to select which model shrinkage to use for that particular prediction.
- Finally we multiply `preds` and `shrinkage` and sum over all levels. 
  - If shrinkage is none, the array will be only zeros and ones, therefore it is equivalent of selecting the model to use
  - If shrinkage is not none, then it is equivalent over averaging with shrinkage factors.